src/adr/0040-generalize-event-bus.md

```diff
@@ -0,0 +1,27 @@
+#40. Generalize EventBus
+
+Date: 2020-06-11
+
+## Status
+
+Proposed
+
+## Context
+
+When searching for users by quota ratio if we set the value of the parameters to 0, the search feature is supposed to return newly created users
+who have not received any email yet at that point. However, this is not the case due to the quotas are currently being initialized only after
+an user receive the first email.
+
+In order to solve this problem, we need to initialize user quotas upon creation time. Event driven system allows indirection so that the users api 
+don't need a dependency on the quota components. The EventBus components currently defined in in mailbox-api would be suitable for this purpose.
+
+## Decision
+
+Extract the EventBus out of mailbox-api in order to make it a utility component so it can be used by all other components.
+
+## Consequences
+
+ - Mailbox-api would leverage the EventBus to keep exposing the MailboxListener api without changes on top of the generified EventBus.  
+ - We can no longer use the sealed hierarchy for events as they depend on different specific components. 
```

Comment:
Let's make the EventBus generic to keep the sealed hierarchy: the hierarchy root could be the parameter of the EventBus

Reply:
Do you mean "having several event buses"? How would we enforce (rabbitmq) resiurce isolation?

What would it bring?

Reply:
> What would it bring?

Keeping the sealed hierarchy which allow to write "total function" listeners (it allows to ensure we don't receive some event types and we don't forget some others)

For example, it allows me to register a `MailboxListener` as today instead of a `Listener` that will just receive `Event` as a type and will finally match only mailbox related events.

> Do you mean "having several event buses"?

It's one of the solutions. We can think about others if this one has problems.

> How would we enforce (rabbitmq) resiurce isolation?

We can define an exchange for each sealed hierarchy.